File: app/src/main/java/com/test/util/accessibility/NodeFinder.java
package com.test.util.accessibility;

import android.os.SystemClock;
import android.view.accessibility.AccessibilityNodeInfo;

import com.test.util.utils.AppLog;

/**
 * 新页面节点加载器...
 * 在即将进入新页面之前调用
 */
public class NodeFinder {

    public static final int perSleepTime = 400;
    public static final int maxLoadTimes = (int) (6000d / perSleepTime);
    private AccessibilityNodeInfo tarNode;
    private String id;
    private String text;
    private int lastPageWindId;
    private boolean isWindIdChange = true; // 目标节点的windID 所在页面是否会改变当前的WinId

    public AccessibilityNodeInfo loadById(String id, boolean isWindIdChange, int lastPageWindId) {
        reset();
        this.id = id;
        this.isWindIdChange = isWindIdChange;
        this.lastPageWindId = lastPageWindId;
        return load();
    }

    public AccessibilityNodeInfo loadByText(String text, boolean isWindIdChange, int lastPageWindId) {
        reset();
        this.text = text;
        this.isWindIdChange = isWindIdChange;
        this.lastPageWindId = lastPageWindId;
        return load();
    }

    public void reset() {
        id = null;
        text = null;
        tarNode = null;
        lastPageWindId = 0;
    }

    private AccessibilityNodeInfo load() {
        for (int i = 0; i < maxLoadTimes; i++) {
            if (isWindIdChange) {
                if (getWinId() != lastPageWindId) {
                    tarNode = getNode();
                }
            } else {
                tarNode = getNode();
            }
            if (tarNode != null) {
                AppLog.INSTANCE.e("找到目标节点：" + getLoadedNodeInfo() + " costTime:" + (maxLoadTimes * perSleepTime));
                return tarNode;
            }
            SystemClock.sleep(perSleepTime);
        }
        AppLog.INSTANCE.e("异常，没有找到目标节点：" + getLoadedNodeInfo() + " costTime:" + (maxLoadTimes * perSleepTime));
        return null;
    }

    private String getLoadedNodeInfo() {
        return " tarNode" + tarNode.hashCode() + " lastPageWindId:" + lastPageWindId;
    }

    public AccessibilityNodeInfo getNode() {
        if (id != null) {
            // TODO
            // AccessibilityUtil.getNodeById(id);
            return null;
        } else if (text != null) {
            // TODO
            // AccessibilityUtil.getNodeByText(text);
            return null;
        } else {
            AppLog.INSTANCE.e("id 和 text不能都是null");
        }
        return null;
    }

    public int getWinId() {
        // TODO
        return 0;
    }
}

File: app/src/main/java/com/test/util/accessibility/lib/NodeFinder.kt
package com.test.util.accessibility.lib

import android.view.accessibility.AccessibilityNodeInfo
import android.os.SystemClock
import com.test.util.utils.AppLog.d
import com.test.util.utils.AppLog.e


/**
 * 节点加载器...
 */
class NodeFinder {

    companion object {
        const val perSleepTime = 300
    }

    private var tarNode: AccessibilityNodeInfo? = null
    private var id: String? = null
    private var text: String? = null
    private var isWinIdChanged = false
    private var lastPageWinId = 0
    private var tarNodeIndex: Int = -1
    private var root: AccessibilityNodeInfo? = null
    private var maxLoadTimes = (6000.0 / perSleepTime).toInt()
    private var nodeCanNull = false

    private var isAutoMode = true

    fun setIsAutoMode(isAutoMode: Boolean){
        this.isAutoMode = isAutoMode
    }

    fun setMaxExeTimes(maxTimes: Int) {
        maxLoadTimes = maxTimes
    }

    fun getNodeById(id: String?, isWinIdChanged: Boolean = false, tarNodeIndex: Int = -1, nodeCanNull: Boolean = false): AccessibilityNodeInfo? {
        reset()
        this.id = id
        this.tarNodeIndex = tarNodeIndex
        this.nodeCanNull = nodeCanNull
        setup(isWinIdChanged)
        return load()
    }

    private fun setup(isWinIdChanged: Boolean) {
        this.isWinIdChanged = isWinIdChanged
        root = ServiceUtil.getService()!!.rootInActiveWindow
        lastPageWinId = root!!.windowId
    }

    fun getNodeByTag(text: String?, isWinIdChanged: Boolean = false, tarNodeIndex: Int = -1, nodeCanNull: Boolean = false): AccessibilityNodeInfo? {
        reset()
        this.text = text
        this.tarNodeIndex = tarNodeIndex
        this.nodeCanNull = nodeCanNull
        setup(isWinIdChanged)
        return load()
    }

    private fun reset() {
        if (isAutoMode && !Start.net.isRunning) {
            throw StopException("stop")
        }
        id = null
        text = null
        tarNode = null
        tarNodeIndex = -1
    }

    private fun load(): AccessibilityNodeInfo? {
        for (i in 0 until maxLoadTimes) {
            if (isWinIdChanged) {
                if (root!!.windowId != lastPageWinId) {
                    tarNode = findNode()
                }
            } else {
                tarNode = findNode()
            }
            if (tarNode != null) {
                d("找到目标节点成功：id:" + id + " text:" + text + " costTime:" + i * perSleepTime)
                return tarNode
            }
            SystemClock.sleep(perSleepTime.toLong())
        }
        if (!nodeCanNull) {
            e("异常，没有找到目标节点：id:" + id + " text:" + text + " costTime:" + (maxLoadTimes * perSleepTime))
        } else {
            d("寻找结束，没有找到节点：id:" + id + " text:" + text + " costTime:" + (maxLoadTimes * perSleepTime))
        }

        return null
    }

    private fun findNode(): AccessibilityNodeInfo? {
        if (id != null) {
            return getNodeById(id!!, false, tarNodeIndex)
        } else if (text != null) {
            return ServiceUtil.getNodeByTag(text!!, tarNodeIndex)
        } else {
            e("id 和 text不能都是null")
        }
        return null
    }

}